File: clox/vm.c
#include <stdarg.h>
#include <stdio.h>
#include <string.h>

#include "common.h"
#include "compiler.h"
#include "vm.h"
#include "debug.h"
#include "object.h"
#include "memory.h"
#include "value.h"

void initVM(VM *vm)
{
    resetStack(vm);
    vm->objects = NULL;
}

void freeVM(VM *vm)
{
    freeObjects(vm);
}

void push(VM *vm, Value value)
{
    *vm->stackTop = value;
    vm->stackTop++;
}

Value pop(VM *vm)
{
    vm->stackTop--;
    return *vm->stackTop;
}
static Value peek(VM *vm, int distance)
{
    return vm->stackTop[-1 - distance];
}

static bool isFalsey(Value value)
{
    return IS_NIL(value) || (IS_BOOL(value) && !AS_BOOL(value));
}

static void concatenate(VM *vm)
{
    ObjString *b = AS_STRING(pop(vm));
    ObjString *a = AS_STRING(pop(vm));

    int length = a->length + b->length;
    char *chars = ALLOCATE(char, length + 1);
    memcpy(chars, a->chars, a->length);
    memcpy(chars + a->length, b->chars, b->length);
    chars[length] = '\0';

    ObjString *result = takeString(vm, chars, length);
    push(vm, OBJ_VAL(result));
}

static void resetStack(VM *vm)
{
    vm->stackTop = vm->stack;
}

static void runtimeError(VM *vm, const char *format, ...)
{
    va_list args;
    va_start(args, format);
    vfprintf(stderr, format, args);
    va_end(args);
    fputs("\n", stderr);

    size_t instruction = vm->ip - vm->chunk->code - 1;
    int line = getLine(vm->chunk, instruction);
    fprintf(stderr, "[line %d] in script\n", line);
    resetStack(vm);
}

InterpretResult interpret(VM *vm, const char *source)
{
    Chunk chunk;
    initChunk(&chunk);

    if (!compile(vm, source, &chunk))
    {
        freeChunk(&chunk);
        return INTERPRET_COMPILE_ERROR;
    }

    vm->chunk = &chunk;
    vm->ip = vm->chunk->code;

    InterpretResult result = run(vm);

    freeChunk(&chunk);
    return result;
}

void negate(VM *vm)
{
    vm->stackTop[-1] = NUMBER_VAL(-AS_NUMBER(vm->stackTop[-1]));
}

static InterpretResult run(VM *vm)
{
#define READ_BYTE() (*vm->ip++)
#define READ_CONSTANT() (vm->chunk->constants.values[READ_BYTE()])
#define BINARY_OP(valueType, op)                                \
    do                                                          \
    {                                                           \
        if (!IS_NUMBER(peek(vm, 0)) || !IS_NUMBER(peek(vm, 1))) \
        {                                                       \
            runtimeError(vm, "Operands must be numbers.");      \
            return INTERPRET_RUNTIME_ERROR;                     \
        }                                                       \
        double b = AS_NUMBER(pop(vm));                          \
        double a = AS_NUMBER(pop(vm));                          \
        push(vm, valueType(a op b));                            \
    } while (false)

    for (;;)
    {
#ifdef DEBUG_TRACE_EXECUTION
        printf("         ");
        for (Value *slot = vm->stack; slot < vm->stackTop; slot++)
        {
            printf("[ ");
            printValue(*slot);
            printf(" ]");
        }
        printf("\n");
        disassembleInstruction(vm->chunk, (int)(vm->ip - vm->chunk->code));
#endif
        uint8_t instruction;
        switch (instruction = READ_BYTE())
        {
        case OP_CONSTANT:
        {
            Value constant = READ_CONSTANT();
            push(vm, constant);
            break;
        }
        case OP_NIL:
            push(vm, NIL_VAL);
            break;
        case OP_FALSE:
            push(vm, BOOL_VAL(false));
            break;
        case OP_TRUE:
            push(vm, BOOL_VAL(true));
            break;
        case OP_EQUAL:
        {
            Value b = pop(vm);
            Value a = pop(vm);
            push(vm, BOOL_VAL(valuesEqual(a, b)));
            break;
        }
        case OP_GREATER:
            BINARY_OP(BOOL_VAL, >);
            break;
        case OP_LESS:
            BINARY_OP(BOOL_VAL, <);
            break;
        case OP_ADD:
        {
            if (IS_STRING(peek(vm, 0)) && IS_STRING(peek(vm, 1)))
            {
                concatenate(vm);
            }
            else if (IS_NUMBER(peek(vm, 0)) && IS_NUMBER(peek(vm, 1)))
            {
                double b = AS_NUMBER(pop(vm));
                double a = AS_NUMBER(pop(vm));
                push(vm, NUMBER_VAL(a + b));
            }
            else
            {
                runtimeError(vm, "Operands must be two numbers or two strings.");
                return INTERPRET_RUNTIME_ERROR;
            }
            // BINARY_OP(NUMBER_VAL, +);
            break;
        }
        case OP_SUBTRACT:
            BINARY_OP(NUMBER_VAL, -);
            break;
        case OP_MULTIPLY:
            BINARY_OP(NUMBER_VAL, *);
            break;
        case OP_DIVIDE:
            BINARY_OP(NUMBER_VAL, /);
            break;
        case OP_NOT:
            push(vm, BOOL_VAL(isFalsey(pop(vm))));
            break;
        case OP_NEGATE:
            if (!IS_NUMBER(peek(vm, 0)))
            {
                runtimeError(vm, "Operand must be a number.");
                return INTERPRET_RUNTIME_ERROR;
            }
            // push(vm, NUMBER_VAL(-AS_NUMBER(pop(vm))));
            negate(vm);
            break;
        case OP_RETURN:
        {
            printValue(pop(vm));
            printf("\n");
            return INTERPRET_OK;
        }
        }
    }

#undef READ_BYTE
#undef READ_CONSTANT
#undef BINARY_OP
}

File: zlox/src/vm.zig
const std = @import("std");
const Chunk = @import("chunk.zig").Chunk;
const OpCode = @import("chunk.zig").OpCode;
const Value = @import("value.zig").Value;
const value = @import("value.zig");
const debug = @import("debug.zig");
const Compiler = @import("compiler.zig").Compiler;

const STACK_MAX = 256;

const InterpretResult = enum {
    OK,
    COMPILE_ERROR,
    RUNTIME_ERROR,
};

pub const VM = struct {
    const Self = @This();
    alloc: std.mem.Allocator,
    chunk: *Chunk = undefined,
    ip: [*]u8 = undefined,
    stack: *[STACK_MAX]Value = undefined,
    stackTop: [*]Value = undefined,

    pub fn init(alloc: std.mem.Allocator) Self {
        var vm = Self{ .alloc = alloc };
        vm.stack = vm.alloc.create([STACK_MAX]Value) catch std.os.exit(1);
        vm.resetStack();
        return vm;
    }

    pub fn deinit(self: *Self) void {
        self.chunk.deinit();
    }

    pub fn interpret(vm: *Self, source: [:0]u8) InterpretResult {
        var chunk = Chunk.init(vm.alloc);
        var compiler = Compiler.init(source, &chunk);
        if (!compiler.compile()) {
            vm.deinit();
            return InterpretResult.COMPILE_ERROR;
        }
        vm.chunk = &chunk;
        vm.ip = vm.chunk.code;
        var result = vm.run();
        return result;
    }

    pub fn run(vm: *Self) InterpretResult {
        while (true) {
            if (debug.trace_enabled) {
                std.debug.print("          ");
                inline for (vm.stack) |slot| {
                    std.debug.print("[ ");
                    value.printValue(slot.*);
                    std.debug.print(" ]");
                }
                std.debug.print("\n", .{});
                debug.disassembleInstruction(vm.chunk, vm.ip - vm.chunk.code.ptr);
            }
            var instruction = @as(OpCode, @enumFromInt(vm.readByte()));
            switch (instruction) {
                .OP_CONSTANT => {
                    var constant = vm.readConstant();
                    vm.push(constant);
                },
                .OP_TRUE => vm.push(Value{ .boolean = true }),
                .OP_FALSE => vm.push(Value{ .boolean = false }),
                .OP_NIL => vm.push(Value{ .nil = undefined }),
                .OP_NEGATE => vm.negate(),
                .OP_ADD => vm.binaryOp(add),
                .OP_SUBTRACT => vm.binaryOp(sub),
                .OP_MULTIPLY => vm.binaryOp(mul),
                .OP_DIVIDE => vm.binaryOp(div),
                .OP_RETURN => {
                    value.printValue(vm.pop());
                    std.debug.print("\n", .{});
                    return InterpretResult.OK;
                },
            }
        }
    }

    inline fn readByte(vm: *Self) u8 {
        var result = vm.ip[0];
        vm.ip += 1;
        return result;
    }

    inline fn readConstant(vm: *Self) Value {
        return vm.chunk.constants.values[readByte(vm)];
    }

    inline fn binaryOp(vm: *Self, comptime op: fn (f64, f64) f64) void {
        var b = vm.pop();
        var a = vm.pop();
        vm.push(Value{ .number = op(a.number, b.number) });
    }

    pub fn resetStack(self: *Self) void {
        self.stackTop = self.stack;
    }

    pub fn push(self: *Self, val: Value) void {
        self.stackTop[0] = val;
        self.stackTop += 1;
    }

    pub fn pop(self: *Self) Value {
        self.stackTop -= 1;
        return self.stackTop[0];
    }

    pub fn negate(self: *Self) void {
        (self.stackTop - 1)[0] = Value{ .number = -(self.stackTop - 1)[0].number };
    }

    fn runtimeError(self: *Self, comptime format: []const u8, args: anytype) void {
        std.debug.panic(format, args);
        std.debug.panic("\n", .{});
        var instruction = self.ip - self.chunk.code - 1;
        const line = self.chunk.getLine(instruction);
        std.debug.panic("[line {}] in script\n", .{line});
        self.resetStack();
    }
};

fn add(a: f64, b: f64) f64 {
    return a + b;
}

fn sub(a: f64, b: f64) f64 {
    return a - b;
}

fn mul(a: f64, b: f64) f64 {
    return a * b;
}

fn div(a: f64, b: f64) f64 {
    return a / b;
}
